FeedStoreChallenge/Core Data/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,121 @@
+//  FeedStoreChallenge
+//
+//  Created by Wilmer Barrios on 23/01/21.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+public final class CoreDataFeedStore: FeedStore {
+	
+	typealias StoredFeedData = (feed: [CDFeedImageItem], timestamp: Date)
+	
+	private let container: NSPersistentContainer
+	private let context: NSManagedObjectContext
+	private let modelName = "FeedStoreDataModel"
+	private let storeURL: URL!
+	
+	public init(storeURL: URL) throws {
+		
+		self.storeURL = storeURL
+		
+		guard let modelURL = Bundle(for: CoreDataFeedStore.self).url(forResource: modelName, withExtension:"momd"),
+			  let model = NSManagedObjectModel(contentsOf: modelURL) else {
+			throw NSError()
+		}
+		
+		let description = NSPersistentStoreDescription(url: storeURL)
+		container = NSPersistentContainer(name: modelName, managedObjectModel: model)
+		container.persistentStoreDescriptions = [description]
+		
+		var loadError: Swift.Error?
+		container.loadPersistentStores { loadError = $1 }
+		try loadError.map { throw $0 }
+		
+		context = container.newBackgroundContext()
+	}
+	
+	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+		let context = self.context
+		context.perform {
+			let fetchRequest: NSFetchRequest<NSFetchRequestResult> = CDFeedImage.fetchRequest()
+			let batchDeleteRequest = NSBatchDeleteRequest(fetchRequest: fetchRequest)
+			do {
+				try context.execute(batchDeleteRequest)
+				completion(nil)
+			} catch let e {
+				completion(e)
+			}
+		}
+	}
+	
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		let context = self.context
+		context.perform {
+			let feed_data: [CDFeedImageItem] = feed.toCD(context: context)
+			let storeFeed = CDFeedImage(context: context)
+			storeFeed.feed_data = NSOrderedSet(array: feed_data)
+			storeFeed.feed_timestamp = timestamp
+			do {
+				try context.save()
+				completion(nil)
+			} catch let e {
+				completion(e)
+			}
+		}
+	}
+	
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		let context = self.context
+		context.perform {
+			let fetchRequest: NSFetchRequest<CDFeedImage> = CDFeedImage.fetchRequest()
+			do {
+				let data = try context.fetch(fetchRequest)
+				let mappedData = self.map(data)
+				guard !mappedData.items.isEmpty else { completion(.empty); return }
+				completion(.found(feed: mappedData.items, timestamp: mappedData.timestamp))
+			} catch let error as NSError {
+				completion(.failure(error))
+				print("Could not fetch. \(error), \(error.userInfo)")
+			}
+		}
+	}
+	
+	// MARK: Helpers
+	private func map(_ data: [CDFeedImage]) -> (items: [LocalFeedImage], timestamp: Date) {
+		var result = [StoredFeedData]()
+		for feed in data {
+			guard let feedData = feed.feed_data, let feedTimestamp = feed.feed_timestamp else { continue }
+			result.append((feedData.array as! [CDFeedImageItem], feedTimestamp))
```

Comment:
Force-casting with `as!` can lead to runtime errors. Instead, it could be part of the guard:

```suggestion
			guard let feedData = feed.feed_data.array as? [CDFeedImageItem], 
			  let feedTimestamp = feed.feed_timestamp else { continue }
			result.append((feedData, feedTimestamp))
```